router/src/queue.rs

```diff
@@ -0,0 +1,142 @@
+use std::cmp::min;
+use crate::InferResponse;
+use crate::{GenerateParameters, GenerateRequest};
+use std::collections::VecDeque;
+use nohash_hasher::IntMap;
+use tokio::sync::mpsc::Receiver;
+use text_generation_client::{
+    Batch, ClientError, NextTokenChooserParameters, Request, StoppingCriteriaParameters,
+};
+use tokio::sync::oneshot::Sender;
+use tokio::time::Instant;
+
+/// In-flight request record
+#[derive(Debug)]
+pub(crate) struct Entry {
+    /// Request
+    pub request: GenerateRequest,
+    /// Response sender to communicate between the Batcher and the batching_task
+    pub response_tx: Sender<Result<InferResponse, ClientError>>,
+    /// Number of tokens in the input
+    pub input_length: usize,
+    /// Instant when this entry was created
+    pub time: Instant,
+    /// Instant when this entry was added to a batch
+    pub batch_time: Option<Instant>,
+}
+
+/// Request Queue
+#[derive(Debug)]
+pub(crate) struct Queue {
+    receiver: Receiver<Entry>,
+    // Staging buffer, filled until max_size is reached
+    buffer: VecDeque<Entry>,
+    /// Id of the next entry
+    next_id: u64,
+    /// Id of the next batch
+    next_batch_id: u64,
+}
+
+
+impl Queue {
+    pub(crate) fn new(receiver: Receiver<Entry>) -> Self {
+        Self { receiver, buffer: VecDeque::new(), next_id: 0, next_batch_id: 0 }
+    }
+
+    /// Get the next batch, blocking until available
+    /// Corresponding entries are added to the entries map
+    /// Returns None only if the queue has been closed
+    pub(crate) async fn next_batch(
+        &mut self,
+        max_size: usize,
+        entries: &mut IntMap<u64, Entry>,
+    ) -> Option<Batch> {
+        if self.buffer.is_empty() {
+            // Await on the queue while the buffer is empty
+            match self.receiver.recv().await {
+                Some(ent) => self.buffer.push_back(ent),
+                // Queue closed, we must be shutting down
+                None => return None,
+            }
+        }
+        // We have at least one entry in the buffer, try to fill it further up to max_size
+        // This will always return Some
+        self.try_next_batch(1, max_size, entries)
+    }
+
+    /// Get the next batch without blocking
+    /// Corresponding entries are added to the entries map
+    pub(crate) fn try_next_batch(
+        &mut self,
+        min_size: usize,
+        max_size: usize,
+        entries: &mut IntMap<u64, Entry>,
+    ) -> Option<Batch> {
+        // Fill the buffer up to max_size, without waiting
+        while self.buffer.len() < max_size {
```

Comment:
I don't understand this part.
If the purpose of this new implem is to optimize the router, why move the enqueue logic on the hot path?

Reply:
I don't understand the question - this loop is calling `try_recv()` to _dequeue_ entries from the channel. They are put into an intermediate buffer only so that we don't have to return them if there are fewer than `min_size`.

Reply:
@OlivierDehaene oh I just realized what a possible source of the confusion is ... I probably should have been clearer that the MPSC channel _is_ the queue.. the VecDequeue is just a kind of staging buffer on the receive side.

Reply:
I know. 
I'm saying if the point is to make things the most async possible, fill this buffer in an async task and either return the buffer directly if it's big enough or None. I don't understand why you fill it when calling `next_batch`. To me this is nullyfing the benefits.
I will write a PR with this logic monday morning to illustrate what I mean.

Reply:
I guess you could have the batcher loop try to fill the buffer while waiting for the shards to return (up to a max of max_batch_size elements), i.e. wait on both the queue and the response future together with `select!`. My intuition is that the time to transfer from the queue via the buffer is negligible and so this wouldn't buy you much, but it could be worth an experiment.

It might also be nice to arrange thing such that the batcher loop is pinned to a dedicated thread... I think this is possible with tokio.